ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala

```diff
@@ -58,6 +59,44 @@ final class SemanticTests extends LedgerTestSuite {
       }
   })
 
+  test(
+    "SemanticConcurrentDoubleSpend",
+    "Cannot concurrently double spend across transactions",
+    allocate(TwoParties, SingleParty),
+  )(implicit ec => {
+    case Participants(Participant(alpha, payer, owner1), Participant(beta, owner2)) =>
+      // This test create a contract and then concurrently archives it several times
+      // through two different participants
+      val contracts = 10 // Number of contracts to create
+      val archives = 10 // Number of concurrent archives per contract
+      Future
+        .traverse((1 to contracts).toList)(c =>
+          for {
+            shared <- alpha.create(payer, SharedContract(payer, owner1, owner2))
+            _ <- synchronize(alpha, beta)
+            results <- Future.traverse((1 to archives).toList)(i =>
+              i % 2 match {
+                case 0 =>
+                  alpha
+                    .exercise(owner1, shared.exerciseSharedContract_Consume1)
+                    .transform(Success(_))
+                case 1 =>
+                  beta
+                    .exercise(owner2, shared.exerciseSharedContract_Consume2)
+                    .transform(Success(_))
+            })
+          } yield {
+            assertLength(s"Contract $c successful archives", 1, results.collect({
+              case Success(_) => true
+            }))
+            assertLength(s"Contract $c failed archives", archives - 1, results.collect({
+              case Failure(_) => true
+            }))
```

Comment:
I think the builtin methods will help you here.

```suggestion
            assertLength(s"Contract $c successful archives", 1, results.filter(_.isSuccess))
            assertLength(s"Contract $c failed archives", archives - 1, results.filter(_.isFailure)))
```